spec/connection.md

```diff
@@ -0,0 +1,258 @@
+
+The flow itself is flipped. TLS is established before the Server sends INFO:
+
+1. Clients initiate a network connection to the Server.
+2. Client upgrades the connection to TLS.
+2. Server [INFO][INFO] json.
+4. Client sends [CONNECT][CONNECT] json.
+5. Clients and Server start to exchange PING/PONG messages to detect if the connection is alive.
+
+
+### Servers discovery
+
+When Server sends back [INFO][INFO]. It may contain additional URLs to which the client can make connection attempts.
+The client should store those URLs and use them in the Reconnection Strategy.
+
+A client should have an option to turn off using advertised URLs.
+By default, those URLs are used.
+
+**TODO**: Add more in-depth explanation how topology discovery works.
+
+### Reconnection Strategies (In progress)
+
+#### Detecting disconnection
+
+There are two methods that clients should use to detect disconnections:
+1. Missing two consecutive PONGs from the Server.
+2. Handling errors from network connection.
+
+#### Reconnect process
+
+When the client detects disconnection, it starts to reconnect attempts with the following rules:
+1. Immediate reconnect attempt
+    The client attempts to reconnect immediately after finding out it has been disconnected.
+2. Exponential BackOff with Jitter
+   - When the first reconnect fails, the backoff process should kick in. Default Jitter should also be included to avoid thundering herd problems.
+3. If the Server returned additional URLs, the client should try reconnecting in random order to each Server on the list.
+4. Successful reconnect resets the timers
+5. Upon reconnection, clients should resubscribe to all created subscriptions.
+
+If there is any change in the connection state - connected/disconnected, the client should have some way of notifying the user about it.
+This can be a callback function or any other idiomatic mechanism in a given language for reporting asynchronous events.
+
+**Disconnect buffer**
+Most clients have a buffer that will aggregate messages on the client side in case of disconnection.
+It will fill up the buffer and send pending messages as soon as connection is restored.
+If buffer will be filled before the conneciton is restored - publish attempts should return error noting that fact.
+
+## Reference-level Explanation
+### Client options
+
+Although clients should provide sensible defaults for handling the connection,
+in many cases, it requires some tweaking.
+The below list defines what can be changed, what it means, and what the defaults are.
+
+#### Ping interval
+
+**default**: 2 minutes
+
+As the client or server might not know that the connection is severed, NATS has Ping/Pong protocol.
+Client can set at what intervals it will send a PING to the server, expecting PONG.
+If two consecutive PONGs are missed, conneciton is marked as lost triggering reconnect attempt.
```

Comment:
Again, not necessarily 2, this is configurable (probably in most client libraries, but maybe not all?).